File: python/blog/sort/quickSort_lomoto.py
from random import *

def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]

def findPivot(arr, le, ri):
    idx = randint(le, ri - 1)
    swap(arr, le, idx)

    return arr[le]

def partition(arr, le, ri):
    pivot = findPivot(arr, le, ri)
    j = k = le

    print('enter partition: ', arr, le, ri)

    for i in range(le + 1, ri):
        print(i, j, k, arr[i], pivot)
        if arr[i] < pivot:
            k += 1
            swap(arr, i, k)
            swap(arr, j, k)
            j += 1
        elif arr[i] == pivot:
            k += 1
            swap(arr, i, k)
    #swap(arr, le, j)

    print('end partition: ', arr, le, ri)
    print('j / k? ', j, k)
    return (j, k)

def quickSort(arr, le, ri):
    if le >= ri:
        return;
    
    j, k = partition(arr, le, ri)

    quickSort(arr, le, j)
    quickSort(arr, k+1, ri)

arr = [15, 20, 13, 30, 10, 25, 3, 24]
#arr = [5, 5, 5, 5, 5]
#arr = [4, 3, 2, 1]
print('before lomoto quick sort: ', arr)
quickSort(arr, 0, len(arr))
print('after lomoto quick sort: ', arr)

File: c++/blog/sort/quickSort_lomoto.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <tuple>

using namespace std;

void print(int* arr, int n) {
	for (int i = 0; i < n; i++)
		cout << *arr++ << " ";
	cout << endl;
}

void swap(int* arr, int i, int j) {
	int tmp = arr[i];
	arr[i] = arr[j];
	arr[j] = tmp;
}

int findPivot(int* arr, int left, int right) {
	srand((unsigned) time(0));
	int i = rand() % (right - left) + left;
	swap(arr, left, i);

	return arr[left];
}

tuple<int, int> partition(int* arr, int left, int right) {
	int pivot = findPivot(arr, left, right);
	int j, k;
	j = k = left;

	for (int i=left+1; i < right; i++) {
		if (arr[i] < pivot) {
			k++;
			swap(arr, i, k);
			swap(arr, j, k);
			j++;
		} else if (arr[i] == pivot) {
			k++;
			swap(arr, i, k);
		}
	}
	
	return make_tuple(j, k);
}

void quickSort(int* arr, int left, int right) {
	if (left >= right)
		return;

	auto result = partition(arr, left, right);
	quickSort(arr, left, get<0>(result));
	quickSort(arr, get<1>(result)+1, right);
}

void main() {
	int arr[] = {15, 20, 13, 30, 10, 25, 3, 24};
	int n = sizeof(arr) / sizeof(*arr);

	cout << "before quick sort: ";
	print(arr, n);

	quickSort(arr, 0, n);

	cout << "after quick sort: ";
	print(arr, n);
}
